### ISSUE TITLE
UX idea: OS + Intention -> Keybinding mappings to Agent

### ISSUE THREAD

**Issue by jbcpollak** [UltimateHackingKeyboard/agent#1201]
How about having a mapping of intentions to hotkeys by OS? Then when configuring a key you could choose the intention/OS combo you want instead of having to remember the key combos for multiple OS.

What I mean is, imagine the agent has a mapping JSON structure like this:

```
{
    "windows10": {
        "next-desktop": "super+ctrl+left-arrow"
    },
    "linux": {
        "next-desktop": "ctrl+pageup"
    }
}
```

When the user selects a key in the agent, add an "Intention" tab to the popup, or add a "find Intention" button to the Keypress tab. There, the user can choose from a drop down (or search) their OS, then choose the intention from another (drop down or search). The selected key sequence gets applies to the key just as if you configured it via the Keypress tab.

In other words, its just a utility to find the key sequence you want, because looking them all up is a pain and for the most part people don't re-map the keys in their OS. The UHK firmware doesn't change because its just a regular keypress, and the agent doesn't need to be running.

As a nice enhancement, maybe the agent could remember that an intention was used to select the keypress, so it could show the user (to help them remember) but maybe that's not possible. Its not critical.

Users could create their own key mappings for lesser used OS's / desktop environments and make PRs back to the agent.

This is a continuation of this thread:

https://twitter.com/joshpollak/status/1240720134177804288?s=20

**Comment by mondalaci** [comment]
I'm sorry, but I still believe that it's fundamentally problematic. There are a number of window managers on Linux for example, and each have their own shortcuts. This would result in confusion and a mainteance burden.

Also, some users would connect their UHKs to different OSes, then wonder why these functions don't work there.

I'm closing this issue, but feel free to follow up regardless.

**Comment by jbcpollak** [comment]
> There are a number of window managers on Linux for example, and each have their own shortcuts. 

This is just like an i18n problem, like "en-us" and "en-gb" - you can have "linux-gnome" and "linux-kde" for example.

> Also, some users would connect their UHKs to different OSes, then wonder why these functions don't work there.

I don't get this - the UHK already ships with layouts for different OSs, this just eases setting those up. It doesn't change the keyboard behavior. You'd still have "QWM" (Mac) and "QWP" (PC) layouts and users would need to switch them when they move between computers.

In fact, I have this exact problem now, and thought of this feature as a fix for it. I just brought my UHK from work (Linux) to my house (Windows) and the key combos do all sorts of different things. So I created a new Windows layout, and I'm in the process of mapping my mod layer to the correct Windows hotkeys. But I have a problem: what Windows hotkeys do the right thing? If the agent had a table I could just look up "windows->next desktop" and apply it to my Windows config. Right now I need to google a mess of hotkeys to get the right combos.

BTW, This proposal isn't a replacement for the regular keypress tab, its just a way to serve 80% of the usecases.

**Comment by mondalaci** [comment]
I can see the usefulness of this feature. But I can also see its drawbacks.

This would result in an ever-growing list of shortcuts in the scancode select2 field, making simple scancodes harder to find, and due to the never-ending shortcut suggestions, it'd pose a long-term mainteance burden.

I respect your opinion and I know you feel strongly about this, but this is not a widely requested feature, and given its cons I can't justify implementing it.

**Comment by jbcpollak** [comment]
Hi @mondalaci - I realize you'd not into this idea, so that's fine, but I wanted to clarify one thing:

> This would result in an ever-growing list of shortcuts in the scancode select2 field, making simple scancodes harder to find, and due to the never-ending shortcut suggestions, it'd pose a long-term mainteance burden.

Agreed it would confusing in the scancode drop down, which is why I'd suggest a separate UI section entirely, either another tab, or a "Find Combo" button that pops up a new control.

I'm not familiar with the Agent architecture, but I don't think this would be a terrible maintenance headache since I think it could be done as data files and separate UI controls.

Maybe someday I'll have a chance to take a crack at it.
